### PR TITLE
[FIX] Multiclass labels do not error when given multilabels

### PR DESCRIPTION
**Pull request by Shivanshmundra**
As per discussion in #2264 and #2262 , this PR solves the issue when multi labels are given to multi-class algorithms where it should raise an error.

Previously it was just taking the label from the first index which was wrong.
Instead of using ```strtol``` function, I tokenized the string and looked for number of labels in an example.

### CODE REVIEW

**Comment by RituRajSingh878** [comment]
you should also add some tests case for the same.

**Comment by jackgerrits** [comment]
The issue raises the problem where the label parser does not fail if the whole token is not consumed. This is not just when accidentally passing a multilabel, it is for any input where the whole token isn't consumed. For example, it should fail for `1a` as well. So the label parser can't just be enlightened to commas.

**Comment by Shivanshmundra** [comment]
So checking whether label string is number or not will check for every case, right?
Something like this function:
```
bool is_number(const std::string& s)  
{  
      std::string::const_iterator it = s.begin();
      while (it != s.end() && std::isdigit(*it)) ++it;
      return !s.empty() && it == s.end();
}

```
But the issue is to get `endptr` to that string, as in this line https://github.com/VowpalWabbit/vowpal_wabbit/blob/e3f4038dbfb890058ff128d69a44ffec7c251236/vowpalwabbit/multiclass.cc#L82 ```words[0]``` contain the whole database as a string and ```endptr``` only return after encountering the first number. 

Should I just check if the string before `|` is a number by using a regular expression or there is some better way?

**Comment by jackgerrits** [comment]
You don't need to do pre-validation. The strtol function return the position in the string that it consumed up to. You will need to modify things so we are able to check if the entire string was consumed or not.

> words[0] contain the whole database as a string and endptr only return after encountering the first number.

Exactly, use the fact that endptr is set after finishing the number conversion.

**Comment by Shivanshmundra** [comment]
@jackgerrits I've changed code to check whether the whole token is consumed or not.
See if this looks fine to you.

File: vowpalwabbit/parse_primitives.h [hunk 1]
```diff
@@ -160,3 +160,22 @@ inline int int_of_string(VW::string_view s)
...
+  if (end <= s.begin() && s.size() > 0)
+  {
+    std::cout << "warning: " << s << " is not a good int, replacing with 0"
+              << std::endl;
+    i = 0;
+  }
+
+  // checking if whole string is consumed
+  if (s.end() != end)
+    return -1;
```
Comment (jackgerrits) [inline vowpalwabbit/parse_primitives.h]: If the user inputted a "-1" how do you distinguish between invalid and valid input?

File: vowpalwabbit/parse_primitives.h [hunk 2]
```diff
@@ -160,3 +160,22 @@ inline int int_of_string(VW::string_view s)
 
   return i;
 }
+
+inline int int_of_string_ml(VW::string_view s)
```
Comment (jackgerrits) [inline vowpalwabbit/parse_primitives.h]: What does `ml` mean?

File: vowpalwabbit/parse_primitives.h (hunk 1 above)
Comment (Shivanshmundra) [reply vowpalwabbit/parse_primitives.h]: Yes, I thought about it but then -1 won't be a label in general. I could resolve this by raising error in the `int_of_string_ml` itself.

File: vowpalwabbit/parse_primitives.h (hunk 2 above)
Comment (Shivanshmundra) [reply vowpalwabbit/parse_primitives.h]: `ml` in `int_of_string_ml` represents multiclass here. As multilabel can have multiple labels, so the error should only invoke in multilabel one. Hence I made another function to get int from the string.

**Comment by Shivanshmundra** [comment]
I changed the name for better interpretability and improved error checking approach to solve the valid and invalid input issues.

**Comment by Shivanshmundra** [comment]
@jackgerrits I was looking at why test cases have been failing from this addition. Actually the issue is with `named_labels`. For example,
 ```
det | the 
noun | monster
```
here label is not an integer but a name. Before this commit, it was giving a warning message ` **label** is not a good float, replacing with 0` which is not correct as it should assign each named label to corresponding integer label.

**Comment by Shivanshmundra** [comment]
@jackgerrits I have rectified all errors. Please have a look.

**Comment by jackgerrits** [comment]
There should not be an implementation of int_of_string that is specialized to this use case. Can you please make it generic through an optional out parameter of the end of the consumed string? That callers can optionally determine whether the whole string was consumed.

**Comment by jackgerrits** [comment]
You have not addressed the concern:
> Can you please make it generic through an optional out parameter of the end of the consumed string?

You have added a bool to check whether the whole token was consumed or not. However, we don't want this function to throw. Please allow the caller to optionally retrieve the endptr that was determined by strtol.

**Comment by Shivanshmundra** [comment]
@jackgerrits Sorry for not getting it right. I still have some confusion regarding your objective.
If I am correct, you want function to return `endptr` of consumed string optionally?

**Comment by jackgerrits** [comment]
Yes, the caller of `int_of_string` should have some mechanism to inspect the value of `endptr` that results from `strtol`

**Comment by Shivanshmundra** [comment]
@jackgerrits One mechanism which I could think of is: 
Since we can't return multiple things in C++ function, declaring a char null pointer before calling function and passing it by reference. It would then store endptr value and then we can do whatever with that ptr.
But this would force to define char variable at every place where `int_of_string` is called which would not be that good idea. Do you have anything good in your mind?

**Comment by jackgerrits** [comment]
You can use an overload to make it optional?

**Comment by jackgerrits** [comment]
Hi @Shivanshmundra. This has been dormant for a while, are you still interested in working on this? If I don't hear from you I am going to go ahead and close it in 7 days.

**Comment by Shivanshmundra** [comment]
Hi @jackgerrits, Actually I was busy in my internship work. Sorry for keeping it dormant.
I couldn't get how to return multiple values in that function. If you can show me any such example which has a similar use case like that of what we are talking about here, I can give a shot.

**Comment by jackgerrits** [comment]
You can effectively return multiple values in C++ using out parameters, references or pointers.

**Pull request closed without merging**